plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,238 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package mavlink
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bluenviron/gomavlib/v3"
+	"github.com/bluenviron/gomavlib/v3/pkg/dialects/ardupilotmega"
+	"github.com/bluenviron/gomavlib/v3/pkg/frame"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type Mavlink struct {
+	URL                    string   `toml:"url"`
+	SystemID               uint8    `toml:"system_id"`
+	FilterPattern          []string `toml:"filter"`
+	StreamRequestFrequency uint16   `toml:"stream_request_frequency"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	filter         filter.Filter
+	connection     *gomavlib.Node
+	endpointConfig gomavlib.EndpointConf
+	cancel         context.CancelFunc
+	wg             sync.WaitGroup
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (*Mavlink) SampleConfig() string {
+	return sampleConfig
+}
+
+func (m *Mavlink) Init() error {
+	// Set default values
+	if m.URL == "" {
+		m.URL = "tcp://127.0.0.1:5760"
+	}
+
+	// Parse out the Mavlink endpoint.
+	// Try to parse the URL
+	u, err := url.Parse(m.URL)
+	if err != nil {
+		return fmt.Errorf("invalid url: %w", err)
+	}
+
+	switch u.Scheme {
+	case "serial":
+		// Serial client
+		// Parse serial URL by hand, because it is not a compliant URL.
+		baudRate := 57600
+		device, rate, found := strings.Cut(strings.TrimPrefix(m.URL, "serial://"), ":")
+		if found {
+			r, err := strconv.Atoi(rate)
+			if err != nil {
+				return fmt.Errorf("serial baud rate not valid: %w", err)
+			}
+			baudRate = r
+		}
+
+		m.endpointConfig = gomavlib.EndpointSerial{
+			Device: device,
+			Baud:   baudRate,
+		}
+	case "tcp":
+		// Use default TCP port if it was not specified
+		port := u.Port()
+		if port == "" {
+			port = "5760"
+		}
+
+		if u.Hostname() == "" {
+			m.endpointConfig = gomavlib.EndpointTCPServer{
+				Address: "0.0.0.0:" + port,
+			}
+		} else {
+			m.endpointConfig = gomavlib.EndpointTCPClient{
+				Address: u.Hostname() + ":" + port,
```

Comment:
Maybe future-proof for IPv6 addresses:
```suggestion
				Address: net.JoinHostPort(u.Hostname(), port),
```